### PR TITLE
feat(server): Add `RestoreSerializer`

### PR DESCRIPTION
**Pull request by chakaz**
This utility class serializes `CompactObj`s as `RESTORE` commands, and has a similar interface (and a common base class) as `RdbSerializer`

### CODE REVIEW

**Review by BorysTheDev — APPROVED** [review]

File: src/server/rdb_save.cc [hunk 1]
```diff
@@ -786,14 +789,21 @@ error_code RdbSerializer::FlushToSink(io::Sink* s) {
...
-  // After every flush we should write the DB index again because the blobs in the channel are
-  // interleaved and multiple savers can correspond to a single writer (in case of single file rdb
-  // snapshot)
-  last_entry_db_index_ = kInvalidDbId;
 
   return error_code{};
 }
 
+error_code RdbSerializer::FlushToSink(io::Sink* s) {
+  error_code res = SerializerBase::FlushToSink(s);
```
Comment (adiholden) [inline src/server/rdb_save.cc:797]: use RETURN_ON_ERR here

File: src/server/rdb_save.cc [hunk 2]
```diff
@@ -1522,4 +1532,37 @@ void RdbSerializer::CompressBlob() {
...
 
+RestoreSerializer::RestoreSerializer(CompressionMode compression_mode)
+    : SerializerBase(compression_mode) {
+}
+
+error_code RestoreSerializer::SaveEntry(const PrimeKey& pk, const PrimeValue& pv,
+                                        uint64_t expire_ms, DbIndex dbid) {
+  absl::InlinedVector<string_view, 4> args;
+
+  string key_buffer;
```
Comment (adiholden) [inline src/server/rdb_save.cc:1543]: you can have key_buffer as a member of this class and save allocations

File: src/server/rdb_save.cc [hunk 3]
```diff
@@ -1522,4 +1532,37 @@ void RdbSerializer::CompressBlob() {
...
+  string_view key = pk.GetSlice(&key_buffer);
+  args.push_back(key);
+
+  string expire_str = absl::StrCat(expire_ms);
+  args.push_back(expire_str);
+
+  string value = RdbSerializer::DumpObject(pv);
```
Comment (adiholden) [inline src/server/rdb_save.cc:1550]: please see how you can change the api of DumpObject to pass it a buffer that you will stor as a member of this class so you wont do so many string allocations while we serialize all entries in database

**Review by adiholden — APPROVED** [review]

**Comment by chakaz** [comment]
PTAL, I fixed a bug that was luckily caught in the unit tests.
And since I was also able to merge https://github.com/romange/helio/pull/201 by now, I also use it to avoid an unneeded string copying on `DUMP` command

**Review by adiholden — APPROVED** [review]

**Comment by chakaz** [comment]
FYI: I gave up on https://github.com/dragonflydb/dragonfly/pull/2366/commits/c3fc8d738196f07aa0d81d5fe0518966a34116a4 because updating Helio seems to break tiered_storage_test, and I'm missing context on debugging that efficiently.

@adiholden You might want to take a look though

**Review by adiholden — APPROVED** [review]

**Pull request merged**